src/metabase/task/sync_databases.clj

```diff
@@ -1,49 +1,169 @@
-            [clojure.tools.logging :as log]
+  "Scheduled tasks for syncing metadata/analyzing and caching FieldValues for connected Databases."
+  (:require [clojure.tools.logging :as log]
             [clojurewerkz.quartzite
+             [conversion :as qc]
              [jobs :as jobs]
              [triggers :as triggers]]
             [clojurewerkz.quartzite.schedule.cron :as cron]
             [metabase
-             [driver :as driver]
-             [sync :as sync]
-             [task :as task]]
+             [task :as task]
+             [util :as u]]
             [metabase.models.database :refer [Database]]
-            [toucan.db :as db]))
-
-(def ^:private ^:const sync-databases-job-key     "metabase.task.sync-databases.job")
-(def ^:private ^:const sync-databases-trigger-key "metabase.task.sync-databases.trigger")
-
-(defonce ^:private sync-databases-job (atom nil))
-(defonce ^:private sync-databases-trigger (atom nil))
-
-;; simple job which looks up all databases and runs a sync on them
-(jobs/defjob SyncDatabases [_]
-  (doseq [database (db/select Database, :is_sample false)] ; skip Sample Dataset DB
-    (try
-      ;; NOTE: this happens synchronously for now to avoid excessive load if there are lots of databases
-      ;; most of the time we do a quick sync and avoid the lengthy analysis process
-      ;; at midnight we run the full sync
-      (let [full-sync? (not (and (zero? (t/hour (t/now)))
-                                 (driver/driver-supports? (driver/engine->driver (:engine database)) :dynamic-schema)))]
-        (sync/sync-database! database {:full-sync? full-sync?}))
-      (catch Throwable e
-        (log/error (format "Error syncing database %d: " (:id database)) e)))))
+            [metabase.sync
+             [analyze :as analyze]
+             [field-values :as field-values]
+             [sync-metadata :as sync-metadata]]
+            [metabase.util.schema :as su]
+            [schema.core :as s]
+            [toucan.db :as db])
+  (:import metabase.models.database.DatabaseInstance
+           [org.quartz CronTrigger JobDetail JobKey TriggerKey]))
+
+;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; |                                                       JOB LOGIC                                                        |
+;;; +------------------------------------------------------------------------------------------------------------------------+
+
+(s/defn ^:private ^:always-validate job-context->database :- DatabaseInstance
+  "Get the Database referred to in JOB-CONTEXT. Guaranteed to return a valid Database."
+  [job-context]
+  (Database (u/get-id (get (qc/from-job-data job-context) "db-id"))))
+
+
+(jobs/defjob SyncAndAnalyzeDatabase [job-context]
+  (let [database (job-context->database job-context)]
+    (sync-metadata/sync-db-metadata! database)
+    ;; only run analysis if this is a "full sync" database
```

Comment:
@salsakran we talked about this last week but we don't need to separate out sync and analysis since analysis only happens now for new fields. We want analysis to happen right away if a new Field is found, but for everything else it's a no-op. So 99.9% of the time this will just run sync and not analysis since there's nothing new to be analyzed.